alibuild_helpers/build.py

```diff
@@ -792,51 +791,60 @@ def downloadTask(p):
+                              spec["version"], args.architecture))),
+        readlink(d))
+      if not match:
+        continue
+      h, revision = match.groups()
+
+      if h != spec["hash"]:
+        if revision.startswith(revisionPrefix) and revision[len(revisionPrefix):].isdigit():
+          # Strip revisionPrefix; the rest is an integer. Convert it to an int
+          # so we can get a sensible max() existing revision below.
+          busyRevisions.add(int(revision[len(revisionPrefix):]))
+        continue
+
+      # Don't re-use local revisions when we have a read-write store, so that
+      # packages we'll upload later don't depend on local revisions.
+      if getattr(syncHelper, "writeStore", False) and "local" in revision:
         continue
-      h, revision = m.groups()
-      revision = int(revision)
 
       # If we have an hash match, we use the old revision for the package
       # and we do not need to build it.
-      if h == spec["hash"]:
-        spec["revision"] = revision
-        if spec["package"] in develPkgs and "incremental_recipe" in spec:
-          spec["obsolete_tarball"] = d
-        else:
-          debug("Package %s with hash %s is already found in %s. Not building.", p, h, d)
-          src = format("%(v)s-%(r)s",
-                       w=workDir,
-                       v=spec["version"],
-                       r=spec["revision"])
-          dst1 = format("%(w)s/%(a)s/%(p)s/latest-%(bf)s",
-                        w=workDir,
-                        a=args.architecture,
-                        p=spec["package"],
-                        bf=spec["build_family"])
-          dst2 = format("%(w)s/%(a)s/%(p)s/latest",
-                        w=workDir,
-                        a=args.architecture,
-                        p=spec["package"])
-
-          getstatusoutput("ln -snf %s %s" % (src, dst1))
-          getstatusoutput("ln -snf %s %s" % (src, dst2))
-          info("Using cached build for %s", p)
-        break
+      spec["revision"] = revision
+      if spec["package"] in develPkgs and "incremental_recipe" in spec:
+        spec["obsolete_tarball"] = d
       else:
-        busyRevisions.append(revision)
-
-    if not "revision" in spec and busyRevisions:
-      spec["revision"] = min(set(range(1, max(busyRevisions)+2)) - set(busyRevisions))
-    elif not "revision" in spec:
-      spec["revision"] = "1"
+        debug("Package %s with hash %s is already found in %s. Not building.",
+              p, h, d)
+        getstatusoutput("ln -snf {ver}-{rev} {wd}/{arch}/{pkg}/latest-{family}".format(
+          ver=spec["version"], rev=spec["revision"], wd=workDir,
+          arch=args.architecture, pkg=spec["package"], family=spec["build_family"]))
+        getstatusoutput("ln -snf {ver}-{rev} {wd}/{arch}/{pkg}/latest".format(
+          ver=spec["version"], rev=spec["revision"], wd=workDir,
+          arch=args.architecture, pkg=spec["package"]))
```

Comment:
Did you change this wrt the current code or is it just a nicer version? Can you do a separate PR for this?

Reply:
No change, just "nicer". I can make a separate PR.

Reply:
Yes, please. We can merge that PR immediately. Given the PR is about a quite critical change, I would like it to be distilled to its minimum.

Reply:
Reverted.